### PR TITLE
Robustness of MVID for reference assemblies

### PR DESCRIPTION
**Pull request by T-Gro**
F# compiler already knows how to produce reference assemblies.
One of the key elements for them is MVID, a deterministic hash identifier used for comparing 'before' and 'after' versions of an assembly.

If MVID of project 'x' is unchanged, projects **depending on** 'x' do not need to recompile and can just copy over the new 'x.dll' without being recompiled themselves. This significantly reduces costs in bigger nets of projects where a change is made in a low-level project, but the change does not alter the visible surface area - e.g. just implementation details are changed, or private members are added.

As of now, the MVID for F# reference assemblies has been calculated both as a hash of IL code, but also as hash of all the embedded resources. For F# that specifically means the Signature and Optimization data (byte streams compiled as resources) used for consumption by other F# projects, but also sections for byte arrays or constant strings for an assembly.

This PR changes the MVID calculation for reference assemblies by not including these resources, and instead replacing them with a custom hash number derived from what I call 'implied signature of a file'.

In a nutshell, it attempts to be equivalent to " auto-generate .fsi signature file and hash that", but without all the intermmediate string allocations and making things pretty - i.e. following the same TAST constructs, just hashing them instead of rendering to a string.

### CODE REVIEW

File: tests/fsharp/Compiler/CodeGen/EmittedIL/DeterministicTests.fs [hunk 1]
```diff
@@ -139,7 +139,7 @@ let test() =
```
Comment (T-Gro) [inline tests/fsharp/Compiler/CodeGen/EmittedIL/DeterministicTests.fs:139]: The body of the tests here is the de-facto **integration test suite** for the ref assembly MVID functionality.
The tests have been here since the ref assembly feature was implemented, most of them were ignore due to the issue this PR addresses.

**Comment by vzarytovskii** [comment]
Does MVID include sig-/opt-data after this change?

I'm trying to figure out whether it's important to take them into account too.

For example - will there be a situation, when we need to re-compile refassembly, and use resources from it, but it won't be recompiled because resources have some info, which is not reflected in the new hash (i.e. it's in resource but not in signature).

**Comment by T-Gro** [comment]
The new MVID does not take resources into account, that is the purpose of it to ensure robustness againts common changes.

If there are changes which are relevant for re-compilation of dependent project, they have to be included in the hash -> that is what I attempted to do by reflecting TAST constructs into it.

**Comment by T-Gro** [comment]
Because if MVID would contain sigdata (like it does in released product), it would not be protected againts hash changes even when private bindings are added, "function within function" gets modified etc.

**Comment by T-Gro** [comment]
> Something to consider including into tests - SRTP function should lead to recompilation when constraint slightly changed.

Yes, there is a long block on the hashing module dedicated to SRTPs.
The only thing which it is robust against is usages of aliases vs. direct types, and order of constrains (the "and" chained list)

**Comment by vzarytovskii** [comment]
> Because if MVID would contain sigdata (like it does in released product), it would not be protected againts hash changes even when private bindings are added, "function within function" gets modified etc.

Yeah, I didn't mean literally hash resources, but data from resources.

**Comment by vzarytovskii** [comment]
Also, what about quotations? We should be hashing them as well.

**Comment by T-Gro** [comment]
> Member

For the dependant project, aren't quatations just a value with Expr or `Expr<T>` ?
Or in which case would we need to consider also the contents of the quotations for the dependent project, and not just the signature?


If anyone can list some examples, I will add it to positive/negative test scenarios (i.e. where hash should or should not change)

**Comment by T-Gro** [comment]
No, the expression itself isn't hashed at the moment, only the name and type is.
So if a low-level project changes from:
`let foo () = <@ 2 + 2 @>`
to
`let foo () = <@ 2 + 3 @>`

It still has the same hash, and higher-level project will not need (as per current hashing) to recompile - but of course AT RUNTIME it will get the freshest .dll and load the .data from it.

Question:
Is there any compile-time scenario for higher-level projects to be aware of full contents of a quotation from a lower-level project?

**Comment by vzarytovskii** [comment]
> Is there any compile-time scenario for higher-level projects to be aware of full contents of a quotation from a lower-level project?

Good question, need to think about it.

**Comment by vzarytovskii** [comment]
> > Is there any compile-time scenario for higher-level projects to be aware of full contents of a quotation from a lower-level project?
> 
> Good question, need to think about it.

I guess it'll only apply if type of the quotation changes, but that will be covered by hashing already.

**Comment by T-Gro** [comment]
> > > Is there any compile-time scenario for higher-level projects to be aware of full contents of a quotation from a lower-level project?
> > 
> > 
> > Good question, need to think about it.
> 
> I guess it'll only apply if type of the quotation changes, but that will be covered by hashing already.

Yes, this I have covered by tests.
Will make robustness tests for `let foo () = <@ 2 + 2 @>` and `let foo () = <@ 2 + 3 @>` having the same hash.

File: src/Compiler/Driver/fsc.fs [hunk 2]
```diff
@@ -842,6 +842,22 @@ let main3
...
+        | MetadataAssemblyGeneration.ReferenceOut _ ->
+            try
+                Fsharp.Compiler.SignatureHash.calculateSignatureHashOfFiles
+                    typedImplFiles
+                    tcGlobals
+                    Fsharp.Compiler.SignatureHash.PublicOnly
+                + Fsharp.Compiler.SignatureHash.calculateHashOfAssemblyTopAttributes topAttrs tcConfig.platform
+                |> Some
+            with e ->
+                printfn "Unexpected error when hashing implied signature, will hash the all of .NET metadata instead. Error: %O " e
```
[the remark this answers is not in the document]
Comment (T-Gro) [reply src/Compiler/Driver/fsc.fs:858]: It's an internal detail, I did not want to have this leaked to the customer - worst case, everything gets hashed as before.
Warning might signalize something is wrong in their code, but it will not be the case...

File: src/Compiler/Service/FSharpCheckerResults.fs [hunk 3]
```diff
@@ -2958,6 +2958,16 @@ type FSharpCheckFileResults
                 |> LayoutRender.showL
                 |> SourceText.ofString)
 
+    member internal _.CalculateSignatureHash() =
+        let visibility = Fsharp.Compiler.SignatureHash.PublicAndInternal
```
[the remark this answers is not in the document]
Comment (T-Gro) [reply src/Compiler/Service/FSharpCheckerResults.fs:2962]: Yes, it means what get's hashed.
Options are either Public or PublicAndInternal.

(this CalculateSignatureHash() was added only for tests)

**Comment by nojaf** [comment]
Hi @T-Gro, very cool to see this work! More stable mvids are absolutely welcome and highly valued.

I noticed that the following scenario is not passing:
```fsharp
[<InlineData("Adding a private let binding",
             (*BEFORE*)"""
module Foo

let a b = b - 1
""",
(*AFTER*)"""
module Foo

let a b = b - 1
let private c = 'd'
""")>]
```

I would expect the mvids to be the same. Would that be a correct assumption or is there still something that ends up in the signature when adding a new private binding?

File: src/Compiler/Driver/fsc.fs (hunk 2 above)
[the remark this answers is not in the document]
Comment (0101) [reply src/Compiler/Driver/fsc.fs:858]: Would be nice if we could send telemetry when this happens. But that's probably out of scope here.

**Review by 0101 — APPROVED** [review]
Looks good! 👍

File: tests/FSharp.Compiler.ComponentTests/Signatures/ImpliedSignatureHashTests.fs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+[<InlineDataAttribute("PrivateModuleAdded",
+(*BEFORE*)"""module MyTest
+type MyRecord = {X:string}"""
+(*AFTER*),"""module MyTest
+type MyRecord = {X:string}
+module private PrivateInnerModule = 
+    let private add a b = a + b""")>]
+
+[<InlineDataAttribute("NestedPrivateModuleAdded",
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Signatures/ImpliedSignatureHashTests.fs:19]:
```suggestion
    let private add a b = a + b""")>]
    
[<InlineData("Adding a private let binding",
(*BEFORE*)"""module Foo
let a b = b - 1
"""
(*AFTER*),"""module Foo
let a b = b - 1
let private c = 'd'   """)>]

[<InlineDataAttribute("NestedPrivateModuleAdded",
```

**Comment by T-Gro** [comment]
> Hi @T-Gro, very cool to see this work! More stable mvids are absolutely welcome and highly valued.
> 
> I noticed that the following scenario is not passing:
> 
> ```fsharp
> [<InlineData("Adding a private let binding",
>              (*BEFORE*)"""
> module Foo
> 
> let a b = b - 1
> """,
> (*AFTER*)"""
> module Foo
> 
> let a b = b - 1
> let private c = 'd'
> """)>]
> ```
> 
> I would expect the mvids to be the same. Would that be a correct assumption or is there still something that ends up in the signature when adding a new private binding?

Thanks for spotting this, commiting it as a (failing) test case and will investigate.
I had a very similar case for private function added, which passes - there must be some subtle difference between the two (most likely the binding is exposed as internal and not private, or something like that)

File: src/Compiler/Driver/fsc.fs (hunk 2 above)
[the remark this answers is not in the document]
Comment (T-Gro) [reply src/Compiler/Driver/fsc.fs:858]: Well this is the fsc.exe, I am not sure we even have telemtry hooked up - we only have it for the VS Editor actions at the moment.

I am very open to any kind of suggestion what to do.
This is the behaviour I want to have:

- It should not a fail a build, falling back to old logic must be preferred over crashing
- If it happens, it is definitely our/my fault -> it must be clear to the person on screen that the issue is not with the code they have written

[the remark this answers is not in the document]
Comment (0101) [reply src/Compiler/Driver/fsc.fs:858]: Yeah that's why I thought telemetry would be great. But don't know what it would take to send it from fcs. 

Maybe when we use this as part of incremental compilation we will be able to send failures through VS telemetry.

**Comment by charlesroddie** [comment]
Fantastic to see progress on this.

Re: private/internal functions, in principle shouldn't both private and internal functions be excluded?

**Comment by T-Gro** [comment]
> Fantastic to see progress on this.
> 
> Re: private/internal functions, in principle shouldn't both private and internal functions be excluded?

If the assembly does not have a single `[<InternalsVisibleToAttribute >]` https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=net-7.0  , impact of internals can be removed too.

But if there is at least one project to which internals are made visible, all internals have to be counted for.

File: tests/FSharp.Compiler.ComponentTests/Signatures/ImpliedSignatureHashTests.fs [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+module private PrivateInnerModule = 
+    let private add a b = a + b""")>]
+    
+[<InlineData("Adding a private let binding",
+(*BEFORE*)"""module Foo
+let a b = b - 1
+"""
+(*AFTER*),"""module Foo
+let a b = b - 1
+let private c = 'd'   """)>]
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Signatures/ImpliedSignatureHashTests.fs:25]: @nojaf  This is the test case from you, in the ``Hash should be stable for`` section == hash is not modified when the private binding for 'c' was added.

**Comment by vzarytovskii** [comment]
@T-Gro mvid calculation should take into account optimization data, otherwise inlining might work incorrectly (not rebuilding ref assembly, when inlined code is changed, but not type) when building against reference assemblies.

**Comment by KevinRansom** [comment]
I also wonder, will the signature files continue to contain the additional unnecessary private data?

**Review by KevinRansom — APPROVED** [review]

**Comment by T-Gro** [comment]
@KevinRansom  ; @vzarytovskii :

Please have a look one more time guys.
Now it does include optimization data and there is a case with an `inline` value to prove that hash changes if it's impl changes.

However, the test suite now relies on the `--nooptimizationdata` flag, which only keeps `MustInline` Vals for optimization, but not the `Optional`. It turns out, typical plain-old-nothing-special bindings are almost always `ValInline.Optional`, which would kind of break purpose of the ref assembly improvements for F#.


Therefore, I will raise a sibling PR with a new approach we discussed with Vlad:

When building in DEBUG && when ref assemblies are present, I propose to build the project with `--nooptimizationdata` . This practically means that only things marked as `inline` really get inlined (including SRTPs), but all the ValInline.Optional Vals are not.


WDYT?

File: src/Compiler/Driver/fsc.fs [hunk 6]
```diff
@@ -850,7 +873,7 @@ let main3
             | Some ib -> ib.RawMetadata.TryGetILModuleDef().Value.MetadataVersion
             | _ -> ""
 
-    let optimizedImpls, optDataResources =
+    let optimizedImpls, (optDataResources, optHash) =
```
Comment (KevinRansom) [inline src/Compiler/Driver/fsc.fs:876]: It's probably better to separate the hash generation from the encodeOptimizationData method.

**Review by KevinRansom — APPROVED** [review]
A minor factoring quibble.  Otherwise it looks okay.

File: src/Compiler/Driver/CompilerImports.fs [hunk 7]
```diff
@@ -220,11 +220,21 @@ let EncodeOptimizationData (tcGlobals, tcConfig: TcConfig, outfile, exportRemapp
             else
                 data
 
+        let ilResource = WriteOptimizationData(tcConfig, tcGlobals, outfile, isIncrementalBuild, ccu, optData)
```
[the remark this answers is not in the document]
Comment (KevinRansom) [reply src/Compiler/Driver/CompilerImports.fs:223]: Oh i'm a twat.

**Review by vzarytovskii — APPROVED** [review]

File: src/Compiler/Driver/fsc.fs (hunk 6 above)
Comment (T-Gro) [reply src/Compiler/Driver/fsc.fs:876]: Will check, did not do it initially because of doubled-processing that would be needed.

Comment (T-Gro) [reply src/Compiler/Driver/fsc.fs]: Done
